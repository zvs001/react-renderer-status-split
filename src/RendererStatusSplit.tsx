import _ from 'lodash'
import React, {ReactElement, ReactNode} from 'react'

interface Props<DataType> {
  statuses: {
    isLoading?: boolean
    isDone?: boolean
    data?: DataType
    error?: string
  }
  isEmpty?: boolean
  renderError?: (error: string) => ReactNode
  renderLoading?: () => ReactNode
  renderPreview?: () => ReactNode
  renderEmpty?: () => ReactNode
  render: (data: this['renderEmpty'] extends undefined ? DataType | undefined : DataType) => ReactNode
}

function RendererStatusSplit<DataType = any>(props: Props<DataType>): ReactElement | null {
  const { statuses, render, renderError, renderLoading, renderEmpty, renderPreview, isEmpty } = props
  const { isLoading, isDone, data, error } = statuses

  let isDataEmpty = _.isEmpty(data)
  if(props.hasOwnProperty('isEmpty')) {
    isDataEmpty = !!isEmpty
  }

  // Early return when data is defined for example by isEmpty prop
  if(!isDataEmpty) { // @ts-ignore
    return render(data)
  }

  if (isLoading && renderLoading) {
    return (<>{renderLoading()}</>)
  }

  if (error && renderError) {
    return <>{renderError(error)}</>
  }

  if (!_.isUndefined(isDone) && !isDone && !_.isUndefined(renderPreview)) {
    if (_.isFunction(renderPreview)) return <>{renderPreview()}</>
    return null
  }



  if (isDataEmpty && renderEmpty) {
    return <>{renderEmpty()}</>
  }

  return <>{render(data as DataType)}</>
}

export default RendererStatusSplit
