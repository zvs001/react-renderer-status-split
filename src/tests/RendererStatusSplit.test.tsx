import ReactTestRenderer from "react-test-renderer";
import React from "react";
import RendererStatusSplit from "../RendererStatusSplit";

const defaultRenderProps = {
  renderPreview: () => <div>preview</div>,
  renderError: (error) => <div>error: {error}</div>,
  renderEmpty: () => <div>empty component</div>,
  renderLoading: () => <div>loading</div>,
  render: () => <div>render data</div>,
};

test("renders default when no handlers", async () => {
  const renderer = ReactTestRenderer.create(
    <RendererStatusSplit
      statuses={{
        isLoading: false,
        data: null,
        error: "",
        isDone: false,
      }}
      render={() => <div>render data</div>}
    />
  );

  expect(renderer).toMatchSnapshot();
});


test("renders preview before statuses change", async () => {
    const renderer = ReactTestRenderer.create(
        <RendererStatusSplit
            statuses={{
                isLoading: false,
                data: null,
                error: "",
                isDone: false,
            }}
            {...defaultRenderProps}
        />
    );

    expect(renderer).toMatchSnapshot();
});

test("isLoading renders correct", async () => {
  const renderer = ReactTestRenderer.create(
    <RendererStatusSplit
      statuses={{
        isLoading: true,
        error: "",
        isDone: false,
      }}
      {...defaultRenderProps}
    />
  );

  expect(renderer).toMatchSnapshot();
});




test("error renders correct", async () => {
    const renderer = ReactTestRenderer.create(
        <RendererStatusSplit
            statuses={{
                isLoading: false,
                error: "unexpected",
                isDone: false,
            }}
            {...defaultRenderProps}
        />
    );

    expect(renderer).toMatchSnapshot();
});

test("isDone renders empty when no data", async () => {
    const renderer = ReactTestRenderer.create(
        <RendererStatusSplit
            statuses={{
                isLoading: false,
                data: null,
                error: "",
                isDone: true,
            }}
            {...defaultRenderProps}
        />
    );

    expect(renderer).toMatchSnapshot();
});


test("isDone renders with data", async () => {
    const renderer = ReactTestRenderer.create(
        <RendererStatusSplit
            statuses={{
                isLoading: false,
                data: '123',
                error: "",
                isDone: true,
            }}
            {...defaultRenderProps}
        />
    );

    expect(renderer).toMatchSnapshot();
});


test("isEmpty(true) is handled correctly", async () => {
    const renderer = ReactTestRenderer.create(
        <RendererStatusSplit
            statuses={{
                isLoading: false,
                error: "",
                isDone: true,
            }}
            isEmpty={true}
            {...defaultRenderProps}
        />
    );

    expect(renderer).toMatchSnapshot();
});


test("isEmpty(false) is handled correctly", async () => {
    const renderer = ReactTestRenderer.create(
        <RendererStatusSplit
            statuses={{
                isLoading: false,
                error: "",
                isDone: true,
            }}
            isEmpty={false}
            {...defaultRenderProps}
        />
    );

    // console.log(renderer.toJSON());
    expect(renderer).toMatchSnapshot();
});



test("isEmpty(false) can skip loading state and proceed to render()", async () => {
    const renderer = ReactTestRenderer.create(
        <RendererStatusSplit
            statuses={{
                isLoading: true,
                error: "",
                isDone: false,
            }}
            isEmpty={false}
            {...defaultRenderProps}
        />
    );

    // console.log(renderer.toJSON());
    expect(renderer).toMatchSnapshot();
});
