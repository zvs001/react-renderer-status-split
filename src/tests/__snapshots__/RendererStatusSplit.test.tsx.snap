// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error renders correct 1`] = `
<div>
  error: 
  unexpected
</div>
`;

exports[`isDone renders empty when no data 1`] = `
<div>
  empty component
</div>
`;

exports[`isDone renders empty when no data 2`] = `
<div>
  render data
</div>
`;

exports[`isDone renders with data 1`] = `
<div>
  render data
</div>
`;

exports[`isEmpty(false) can skip loading state and proceed to render() 1`] = `
<div>
  render data
</div>
`;

exports[`isEmpty(false) is handled correctly 1`] = `
<div>
  render data
</div>
`;

exports[`isEmpty(true) is handled correctly 1`] = `
<div>
  empty component
</div>
`;

exports[`isLoading renders correct 1`] = `
<div>
  loading
</div>
`;

exports[`renders default when no handlers 1`] = `
<div>
  render data
</div>
`;

exports[`renders preview before statuses change 1`] = `
<div>
  preview
</div>
`;
